specs/capella/beacon-chain.md

```diff
@@ -282,6 +289,19 @@ def is_fully_withdrawable_validator(validator: Validator, epoch: Epoch) -> bool:
     return is_eth1_withdrawal_prefix and validator.withdrawable_epoch <= epoch < validator.fully_withdrawn_epoch
 ```
 
+#### `is_partially_withdrawable_validator`
+
+```python
+def is_partially_withdrawable_validator(validator: Validator, balance: Gwei) -> bool:
+    """
+    Check if ``validator`` is partially withdrawable.
+    """
+    is_eth1_withdrawal_prefix = validator.withdrawal_credentials[:1] == ETH1_ADDRESS_WITHDRAWAL_PREFIX
+    has_max_effective_balance = validator.effective_balance == MAX_EFFECTIVE_BALANCE
+    has_excess_balance = balance > MAX_EFFECTIVE_BALANCE
+    return is_eth1_withdrawal_prefix and has_max_effective_balance and has_excess_balance
```

Comment:
Shouldn't it also check that the validator is in active state?

Reply:
> Shouldn't it also check that the validator is in active state?

I think it would be reasonable for an exiting validator, for example, to be able to skim.  May want to consider if slashed validators can do so, however.

Reply:
My main concern is related to slashed validators. Also, i don't see value of skimming exited validators as their next step should be full withdrawal, it feels like unnecessary overhead in the withdrawal churn.

Reply:
Exited is different from exiting, however.  A validator could theoretically be stuck in exiting for quite a while, if a lot of validators head for the door at the same time so we have a large exit queue.

Although perhaps this is just a case of our definitions of "active" not being quite the same, and we're actually saying the same thing.  For me, I think that `validator.slashed == false && validator.exit_epoch > current_epoch" would be a sufficient condition (and yes this would theoretically catch validators before they activated, but I don't see that being a problem as they will only have excess balance if there was a double deposit and no harm in that coming back to the user as early as possible).

Reply:
I don't really see an issue with slashed or not-yet-activated validators getting a partial withdrawal about 32 ETH.

* not-yet-activated && balance about 32 -- will be partially withdrawn immediately upon activation
* slashed -- they cannot lose more than 32 eth (in most cases) so any excess will become withdrawn anyway
* exited-but-not-withdrawn -- restricting this just adds additional complexity and testing on this condition for near zero value. the amount of load induced by partial withdrawals is bound and will almost certainly always use the max-per-epoch (assuming a minimum number of 0x01 credentials). So the cost on the system of these types of withdrawals is 0 and the impact on limiting other validators access to partial withdrawals is low.

I'm not fundamentally opposed to the above conditions being added, but I think they provide little value.

*Note*: the way a validator can lose more than 32 ETH via slashing is if they have a highly correlated fault (so high base slashing, e.g. 32 ETH) AND they suffer a high inactivity leak prior to the 32 ETH slashing.

Reply:
Thinking it through, a slashed validator will have lost `balance/MIN_SLASHING_PENALTY_QUOTIENT` when it enters the slashing state so although I think that philosophically speaking a slashed validator shouldn't be able to obtain any withdrawals from the chain whilst it is being slashed, in practical terms this will never happen anyway (outside of a couple of edge cases that I don't think are worthwhile worrying about) so you are correct no need to codify it here.

Regarding `validator.exit_epoch > current_epoch`: this came from the thinking that partial withdrawals were for validators that were validating, rather than those that have exited and will be fully withdrawn "soon" anyway.  I'd prefer to see it in there, but not so much that I'd fight for its inclusion, and if it creates additional complexity elsewhere then fair enough to not have it present.

Reply:
I'm cleaning up this PR now and after coming back to it, I think I agree that this is actually simpler to reason about and test by doing just active (and not slashed) rather than all potential validators with excess balances.

I'm going to alter the PR and tests to that and we can touch on the discussion once more

Reply:
I just went through the motions and it did add moderate complexity in testing to condition things upon epoch rather than purely on balance.

I want to get this merged and will open up an issue for discussion if anyone feels strongly about this